onnxruntime/core/providers/cpu/math/einsum_utils/einsum_typed_compute_processor.cc

```diff
@@ -0,0 +1,359 @@
+  if (EinsumOp::IsTransposeRequired(candidate_output_shape_without_reduced_dims.size(), output_permutation)) {
+    auto candidate_output_transposed = EinsumOp::Transpose(candidate_output, candidate_output_shape_without_reduced_dims,
+                                                           output_permutation,
+                                                           allocator_, cublas_handle_, device_transpose_func_);
+
+    // We have an the result in an output "candidate". Now we have to copy the contents in its buffer
+    // into the buffer of the actual output given to us by the execution frame
+    // We need to do this because the buffer owned by the output tensor of the op could be user provided buffer
+
+    auto status = device_data_copy_func_(*candidate_output_transposed, output);
+    ORT_ENFORCE(status.IsOK(), "Einsum op: Could not copy the intermediate output's buffer into the op's output buffer. Error: ",
+                status.ErrorMessage());
+
+  } else {
+    // Copy the output candidate into the op's output
+    auto status = device_data_copy_func_(candidate_output, output);
+    ORT_ENFORCE(status.IsOK(), "Einsum op: Could not copy the intermediate output's buffer into the op's output buffer. Error: ",
+                status.ErrorMessage());
+  }
+}
+
+template <typename T>
+std::unique_ptr<Tensor> EinsumTypedComputeProcessor<T>::PairwiseOperandProcess(const Tensor& left,
+                                                                               const TensorShape& left_shape_override,
+                                                                               const Tensor& right,
+                                                                               const TensorShape& right_shape_override,
+                                                                               const std::vector<int64_t>& reduce_dims,
+                                                                               bool is_final_pair) {
+  // Use the provided dim overrides instead of the actual shapes of the operands
+  ORT_ENFORCE(left.Shape().Size() == left_shape_override.Size(),
+              "The override dims are not compatible with given tensor's shape");
+  ORT_ENFORCE(right.Shape().Size() == right_shape_override.Size(),
+              "The override dims are not compatible with given tensor's shape");
+
+  // Make copy as this may be overridden downstream
+  const auto& left_dims = left_shape_override.GetDims();
+  const auto& right_dims = right_shape_override.GetDims();
+
+  int64_t left_rank = static_cast<int64_t>(left_dims.size());
+  int64_t right_rank = static_cast<int64_t>(right_dims.size());
+
+  std::unique_ptr<Tensor> current_left;
+  std::unique_ptr<Tensor> current_right;
+
+  // If the following error condition is hit, it is most likely a pre-processing bug
+  ORT_ENFORCE(left_rank == right_rank, "Ranks of pair-wise operands must be equal");
+
```

Comment:
nit: would be good to provide the shapes in the output so a user can see what wasn't equal

Reply:
added it here and a couple of other places.